File: framework/ndn_iot_cpp/examples/test-producing.cpp
#include "bootstrap.hpp"
#include "common.hpp"
#include "boost/asio.hpp"

using namespace ndn;
using namespace std;

namespace ndn_iot {

namespace examples {

class AppConsumer
{
public:
  AppConsumer() {
    face_ = ptr_lib::shared_ptr<ThreadsafeFace>(new ThreadsafeFace(ioService_));
  }

  void run() {
    Bootstrap bootstrap(*(face_.get()));
    keyChain_ = bootstrap.setupDefaultIdentityAndRoot(Name("/home/my-cpp-device"));
    defaultCertificateName_ = bootstrap.getDefaultCertificateName();
    cout << "default certificate name is: " << defaultCertificateName_.toUri() << endl;
    memoryContentCache_.reset(new MemoryContentCache(face_.get()));
    memoryContentCache_->registerPrefix(Name("/home/flow/cpp-publisher-1"), 
      bind(&AppConsumer::onRegisterFailed, this, _1));

    bootstrap.requestProducerAuthorization(Name("/home/flow/cpp-publisher-1"), "flow",
      bind(&AppConsumer::onRequestSuccess, this), 
      bind(&AppConsumer::onRequestFailed, this, _1));
    boost::asio::io_service::work work(ioService_);
    ioService_.run();
  }
private:
  void onRequestSuccess() {
    cout << "request granted" << endl;
    Data data(Name("/home/flow/cpp-publisher-1/0"));
    string content = "good good";
    data.setContent((const uint8_t *)&content[0], content.size());
    keyChain_->sign(data, defaultCertificateName_);

    memoryContentCache_->add(data);
  }

  void onRequestFailed(string msg) {
    cout << "request not granted: " << msg << "; in this test, publish anyway" << endl;

    Data data(Name("/home/flow/cpp-publisher-1/0"));
    string content = "good good";
    data.setContent((const uint8_t *)&content[0], content.size());
    keyChain_->sign(data, defaultCertificateName_);

    memoryContentCache_->add(data);
  }

  void onVerified(const ptr_lib::shared_ptr<const Data>& data) {
    cout << "data verified: " << data->getName().toUri() << endl;
  }

  void onVerifyFailed(const ptr_lib::shared_ptr<const Data>& data) {
    cout << "data verification failed: " << data->getName().toUri() << endl;
  }

  void onRegisterFailed(const ptr_lib::shared_ptr<const Name>& prefix) {
    cout << "Prefix registration failed: " << prefix->toUri() << endl;
  }

  ptr_lib::shared_ptr<ThreadsafeFace> face_;
  boost::asio::io_service ioService_;
  ptr_lib::shared_ptr<KeyChain> keyChain_;
  ptr_lib::shared_ptr<MemoryContentCache> memoryContentCache_;
  Name defaultCertificateName_;
};

}

}

int main() {
  ndn_iot::examples::AppConsumer consumer;
  consumer.run();

  return 0;
}

File: framework/ndn_iot_dot_net/examples/test-producing.cs
namespace ndn_iot.tests {
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using System.IO;

    // TODO: to be finished up with consumer functions as in other languages
    using System.Runtime.InteropServices;

    using net.named_data.jndn.security.policy;    
    using net.named_data.jndn;
    using net.named_data.jndn.encoding;
    using net.named_data.jndn.util;
    using net.named_data.jndn.security;
    using net.named_data.jndn.security.identity;
    using net.named_data.jndn.security.certificate;
    using net.named_data.jndn.encoding.tlv;

    using net.named_data.jndn.transport;

    using ndn_iot.bootstrap;

    class AppProducer : OnRegisterFailed, OnInterestCallback, OnRegisterSuccess {
        class DummyHandler : OnData, OnTimeout {
            public DummyHandler(AppProducer apProducer) {
                apProducer_ = apProducer;
            }

            public void onData(Interest interest, Data data) {
                Console.Out.WriteLine("Got dummy data, unexpected");
            }

            public void onTimeout(Interest interest) {
                apProducer_.publishData();
            }

            AppProducer apProducer_;
        }

        public AppProducer(Name producerNamespace) {
            producerNamespace_ = producerNamespace;
            dh_ = new DummyHandler(this);
        }

        public void publishData() {
            TimeSpan t = DateTime.UtcNow - new DateTime(1970, 1, 1);

            Data data = new Data(new Name(producerNamespace_).append(Name.Component.fromVersion((long)t.TotalSeconds)));
            string content = "good good";
            data.setContent(new Blob(content));
            keyChain_.sign(data, certificateName_);

            memoryContentCache_.add(data);
            Console.Out.WriteLine("Data published: " + data.getName().toUri());

            Interest dummyInterest = new Interest(new Name("/local/timeout"));
            dummyInterest.setInterestLifetimeMilliseconds(4000);
            face_.expressInterest(dummyInterest, dh_, dh_);
        }

        void onRequestSuccess() {
            Console.Out.WriteLine("Request granted!");
            publishData();
        }

        void onRequestFailed(string msg) {
            Console.Out.WriteLine(msg);
            Console.Out.WriteLine("Request failed, but in this test start publishing anyway");
            publishData();
        }

        public void run()
        {
            face_ = new Face(new TcpTransport(), new TcpTransport.ConnectionInfo("localhost"));
            Bootstrap bootstrap = new Bootstrap(face_);
            keyChain_ = bootstrap.setupDefaultIdentityAndRoot(new Name("/home/flow-csharp"), new Name());
            certificateName_ = bootstrap.getDefaultCertificateName();
            memoryContentCache_ = new MemoryContentCache(face_);
            
            // separate debug function for creating ID and cert
            //bootstrap.createIdentityAndCertificate(new Name("/home/flow/csharp-publisher-1"));

            // main is static so cannot refer to non-static members here, if want to make onRequestSuccess and onRequestFailed non-static
            bootstrap.requestProducerAuthorization(
              new Name(producerNamespace_), 
              "flow", 
              new OnRequestSuccess(onRequestSuccess), 
              new OnRequestFailed(onRequestFailed));

            memoryContentCache_.registerPrefix(producerNamespace_, this, this, this);

            while (true) {
                face_.processEvents();
                // We need to sleep for a few milliseconds so we don't use 100% of the CPU.
                System.Threading.Thread.Sleep(5);
            }
        }

        public void onRegisterSuccess(Name prefix, long id) {
            Console.Out.WriteLine("Prefix registration success");
        }

        public void onRegisterFailed(Name prefix) {
            Console.Out.WriteLine("Prefix registration failed for: " + prefix.toUri());
        }

        public void onInterest(Name prefix, Interest interest, Face face, long interestFilterId, InterestFilter filter) {
            Console.Out.WriteLine("Data not found for interest: " + interest.getName().toUri());
        }

        public KeyChain keyChain_;
        public Face face_; 
        public Name certificateName_;
        public MemoryContentCache memoryContentCache_;

        private Name producerNamespace_;
        private DummyHandler dh_;
    }

    class TestProducer {
        static void Main(string[] args) {
            AppProducer apProducer = new AppProducer(new Name("/home/flow/ts-publisher-1"));
            apProducer.run();
        }
    }
}